src/transformStateWithClones.js

```diff
@@ -7,7 +7,48 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
-}
+  const stateHistory = [];
+  let stateCopy = { ...state };
+
+  const addProperties = (extraData) => {
+    stateCopy = Object.assign(stateCopy, extraData);
+    stateHistory.push({ ...stateCopy });
+  };
+
+  const removeProperties = (keysToRemove) => {
+    for (const remove in keysToRemove) {
+      for (const keys in stateCopy) {
+        if (keysToRemove[remove] === keys) {
+          delete stateCopy[keys];
+        }
+      }
+    }
+
+    stateHistory.push({ ...stateCopy });
+  };
+
+  const clear = () => {
+    for (const keys in stateCopy) {
```

Comment:
Instead of deleting all properties one by one, you could simply assign an empty object to 'stateCopy' to clear it. This would be more efficient and clearer to read.